stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,257 @@ public extension Tensor where Scalar : Numeric {
+  @differentiable(wrt: self, vjp: _vjpSlice)
+  func slice(begin: Tensor<Int32>, size: Tensor<Int32>) -> Tensor {
+    return Raw.slice(self, begin: begin, size: size)
+  }
+
+  @inlinable @inline(__always)
+  internal func _vjpSlice(
+    begin: Tensor<Int32>,
+    size: Tensor<Int32>
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    let value = slice(begin: begin, size: size)
+    let afterPaddings = shapeTensor - value.shapeTensor - begin
+    return (value, { [after = afterPaddings] v in
+      let beforePaddings = begin.expandingShape(at: 1)
+      let afterPaddings = after.expandingShape(at: 1)
+      let paddings = Tensor<Int32>(concatenating: [beforePaddings, afterPaddings], alongAxis: 1)
+      return Raw.pad(v, paddings: paddings)
+    })
+  }
+}
+
+public enum TensorSliceIndex : TensorSliceIndexProtocol {
+  case ellipsis
+  case newAxis
+  case squeezeAxis
+  case index(Int32)
+  case range(Range<Int32>, stride: Int32)
+  case closedRange(ClosedRange<Int32>, stride: Int32)
+  case partialRangeFrom(PartialRangeFrom<Int32>, stride: Int32)
+  case partialRangeUpTo(PartialRangeUpTo<Int32>, stride: Int32)
+  case partialRangeThrough(PartialRangeThrough<Int32>, stride: Int32)
+
+  public var sliceIndex: TensorSliceIndex { return self }
+}
+
+public protocol TensorSliceIndexProtocol {
+  var sliceIndex: TensorSliceIndex { get }
+}
+
+// TODO: Cannot extend non-nominal type 'UnboundedRange'.
+// extension UnboundedRange : TensorSliceIndexProtocol {
+//   public var sliceIndex: TensorSliceIndex { return .ellipsis }
+// }
+
+extension Int32 : TensorSliceIndexProtocol {
+  public var sliceIndex: TensorSliceIndex { return .index(self) }
+}
+
+extension Int : TensorSliceIndexProtocol {
+  public var sliceIndex: TensorSliceIndex { return .index(Int32(self)) }
+}
+
+extension Range : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .range(Int32(self.lowerBound)..<Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension ClosedRange : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .closedRange(Int32(self.lowerBound)...Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeFrom : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeFrom(Int32(self.lowerBound)..., stride: 1)
+  }
+}
+
+extension PartialRangeUpTo : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeUpTo(..<Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeThrough : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeThrough(...Int32(self.upperBound), stride: 1)
+  }
+}
+
+public extension Tensor {
+  public struct IndexPath {
+    public let begin, end, strides: Tensor<Int32>
+    public let beginMask, endMask, ellipsisMask, newAxisMask, squeezeAxisMask: Int64
```

Comment:
Is there a reason these are implemented as integer masks? What if the index is greater than 64?

Especially when we need to expose them as public properties, they should be general types instead of masks. An array of integers feels like a better solution.

Reply:
Yes, the main reason is that the strided slice is implemented using these masks in TensorFlow and I am using this struct as the intermediate representation that's then passed directly to the TF op. We could make these fields internal actually since I don't see any reason why they should be exposed to the users.

Reply:
Internal sounds good.